design/ingressroute-design.md

```diff
@@ -0,0 +1,300 @@
+
+# Non-goals
+
+- Support ingress objects outside the current cluster
+- IP in IP or GRE tunneling between custers with discontinious or overlapping IP ranges.
+- Support non HTTP/HTTPS traffic ingress (ie, no UDP, no MongoDB, no TCP passthrough). We just do HTTP/HTTPS ingress.
+- Deprecate Contours support for v1beta1.Ingress.
+
+# Background
+
+The ingress object was added to kubernetes in 1.2 as a way of describing properties of a cluster wide reverse HTTP proxy.
+Since that time the ingress object has not progressed beyond the beta stage, and its stagnation inspired a cambrian explosion of annotations to express missing properties of http routing.
+
+# High-Level Design
+
+At a high level, this document proposes modeling ingress configuation as a graph of documents throughout a Kubernetes API server which when taken together form a directed acyclic graph (DAG) of the configuration for virtual hosts and their constituent routes.
+
+## Delegation
+
+The working model for delegation is DNS.
+As the owner of a DNS domain, for example `.com`, I _delegate_ to another nameserver the responsibility for handing the subdomain `heptio.com`.
+Any nameserver can hold a record for `heptio.com`, but without the linkage from the parent `.com` TLD, its information is unreachable and non authorative.
+
+Each _root_ of a DAG starts at a virtual host, which describes properties like the fully qualified name of the virtual host, any aliases (for example a www. prefix) of that vhost, TLS configuration, and possibly global access list details.
+The edges of a graph do not contain virtual host information, they are only reachable from a root via a delegation.
+This permits the _owner_ of an ingress root to both delegate the authority to publish a service on a portion of the route space inside a virtual host, and to itself further delegate authority to publish and delegate.
+
+This also means that 
+
+In practice the linkage, or delegation, from root to edge, is performed with a specific type of route action.
+You can think of it that rather than routing traffic to a service, it is routed to another ingressroute object for further processing.
+
+
+##
+
+# Detailed Design
+
+## IngressRoute CRD
+
+This is an example of a fully populated root ingressroute.
+
+```yaml
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  name: google
+  namespace: prod
+spec:
+  # virtualhost appears at most once. If it is present, the object is considered
+  # to be a "root".
+  virtualhost:
+    # the fully qualified domain name of the root of the ingress tree
+    # all leaves of the DAG rooted at this object relate to the fqdn (and its aliases)
+    fqdn: www.google.com
+    # a set of aliases for the domain, these may be alternative fqdns which are considered
+    # aliases of the primary fqdn
+    aliases:
```

Comment:
How would wildcard domains work?

Reply:
I'm super leary of wildcard domains. They are a total hack in the current ingress spec. The ingress spec says the host: field must be a fully qualified domain as defined by (some spec I can't remember atm), and * is not one of the allowed characters.

The use case of wildcard domains is user generated subdomains all routed through to a single web application, say heptio.tumblr.com. 

The solution to this might be change the name of the `fqdn` field to something like `name` or `host` to move it closer to "this is what the browser is going to tell you", not "this is what you configured in dns".

Reply:
We need a plan around how to support, I've used `*.domain.com` many many times in the past and had a CI system write Ingress records to wire up subdomains to apps utilizing a wildcard cert on the front.

Reply:
@jbeda excellent suggestion, I think the model of "this configuration applies to these (n) fqdns" is valid, but making one the "primary" and others aliases is of little value. 

However, this is contingent on the security issue I raised at the end of the document. The goal of having a "primary" domain name was to avoid having to list all the aliases at each level of the delegation chain.